File: packages/kt-client/src/main/kotlin/dev/pyric/database/OnDisconnect.kt
package dev.pyric.database

import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.TaskCompletionSource
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class OnDisconnect(private val ref: DatabaseReference) {
    private val scope = CoroutineScope(Dispatchers.IO)

    fun setValue(value: Any?): Task<Void?> = executeOp(
        method = "rtdb.onDisconnectSet",
        params = mapOf("path" to ref.path, "value" to value)
    )

    fun setValue(value: Any?, priority: Any?): Task<Void?> = executeOp(
        method = "rtdb.onDisconnectSet",
        params = mapOf("path" to ref.path, "value" to value, "priority" to priority)
    )

    fun updateChildren(update: Map<String, Any?>): Task<Void?> = executeOp(
        method = "rtdb.onDisconnectUpdate",
        params = mapOf("path" to ref.path, "values" to update)
    )

    fun removeValue(): Task<Void?> = executeOp(
        method = "rtdb.onDisconnectRemove",
        params = mapOf("path" to ref.path)
    )

    fun cancel(): Task<Void?> = executeOp(
        method = "rtdb.onDisconnectCancel",
        params = mapOf("path" to ref.path)
    )

    private fun executeOp(method: String, params: Map<String, Any?>): Task<Void?> {
        val tcs = TaskCompletionSource<Void?>()
        scope.launch {
            try {
                ref.database.bridgeClient.op(
                    method = method,
                    params = params,
                    actAs = ref.database.getEffectiveAuthLens().toMap()
                )
                tcs.setResult(null)
            } catch (e: Exception) {
                tcs.setException(e)
            }
        }
        return tcs.task
    }
}

File: packages/swift-client/Sources/PyricDatabase/OnDisconnect.swift
import Foundation
import PyricFirestore

/// Registers server-side operations to execute when the client disconnects.
public final class OnDisconnect: NSObject, @unchecked Sendable {
    public let ref: DatabaseReference

    public init(ref: DatabaseReference) {
        self.ref = ref
        super.init()
    }

    public func setValue(_ value: Any?) async throws {
        let params: [String: AnySendable] = [
            "path": .string(ref.path),
            "value": RtdbValueCodec.toAnySendable(value)
        ]
        _ = try await ref.database.bridgeClient.op(
            method: "rtdb.onDisconnectSet",
            params: params,
            actAs: ref.database.currentAuthLens()
        )
    }

    public func setValue(_ value: Any?, andPriority priority: Any?) async throws {
        var params: [String: AnySendable] = [
            "path": .string(ref.path),
            "value": RtdbValueCodec.toAnySendable(value)
        ]
        if let priority {
            params["priority"] = RtdbValueCodec.toAnySendable(priority)
        }
        _ = try await ref.database.bridgeClient.op(
            method: "rtdb.onDisconnectSet",
            params: params,
            actAs: ref.database.currentAuthLens()
        )
    }

    public func updateChildValues(_ values: [AnyHashable: Any]) async throws {
        let params: [String: AnySendable] = [
            "path": .string(ref.path),
            "values": RtdbValueCodec.toAnySendable(values)
        ]
        _ = try await ref.database.bridgeClient.op(
            method: "rtdb.onDisconnectUpdate",
            params: params,
            actAs: ref.database.currentAuthLens()
        )
    }

    public func removeValue() async throws {
        let params: [String: AnySendable] = [
            "path": .string(ref.path)
        ]
        _ = try await ref.database.bridgeClient.op(
            method: "rtdb.onDisconnectRemove",
            params: params,
            actAs: ref.database.currentAuthLens()
        )
    }

    public func cancel() async throws {
        let params: [String: AnySendable] = [
            "path": .string(ref.path)
        ]
        _ = try await ref.database.bridgeClient.op(
            method: "rtdb.onDisconnectCancel",
            params: params,
            actAs: ref.database.currentAuthLens()
        )
    }
}
